README.md
# Time-based One-Time Password Lib _(@digitalbazaar/totp)_

[![Build status](https://img.shields.io/github/workflow/status/digitalbazaar/totp/Node.js%20CI)](https://github.com/digitalbazaar/totp/actions?query=workflow%3A%22Node.js+CI%22)
[![Coverage status](https://img.shields.io/codecov/c/github/digitalbazaar/totp)](https://codecov.io/gh/digitalbazaar/totp)
[![NPM Version](https://img.shields.io/npm/v/@digitalbazaar/totp.svg)](https://npm.im/@digitalbazaar/totp)

> A library with Time-based One-Time Password (TOTP) utilities that run in
Node.js and in the browser.

## Table of Contents

- [Background](#background)
- [Security](#security)
- [Install](#install)
- [Usage](#usage)
- [Contribute](#contribute)
- [Commercial Support](#commercial-support)
- [License](#license)

## Background

TBD

## Security

TBD

## Install

- Node.js 14+ is required.

### NPM

To install via NPM:

```
npm install @digitalbazaar/totp
```

### Development

To install locally (for development):

```
git clone https://github.com/digitalbazaar/totp
cd totp
npm install
```

## Usage

TBD

## Contribute

See [the contribute file](https://github.com/digitalbazaar/bedrock/blob/master/CONTRIBUTING.md)!

PRs accepted.

If editing the Readme, please conform to the
[standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## Commercial Support

Commercial support for this library is available upon request from
Digital Bazaar: support@digitalbazaar.com

## License

[New BSD License (3-clause)](LICENSE) © Digital Bazaar


CHANGELOG.md
# totp Changelog

## 1.0.1 - 2022-05-22

### Fixed
- Ensure `base32Decode` returns a `Uint8Array`.

## 1.0.0 - 2022-05-22

### Added
- Initial implementation.


tests/benchmark.js
/*!
* Copyright (c) 2020-2022 Digital Bazaar, Inc. All rights reserved.
*/
import Benchmark from 'benchmark';

const suite = new Benchmark.Suite();

import {
  generateSecret, generateToken
} from '../lib/index.js';

suite
  .add('generateToken', {
    defer: true,
    fn: async deferred => {
      // FIXME:
      const params = await generateSecret();
      await generateToken(params);
      deferred.resolve();
    }
  })
  .on('cycle', event => {
    console.log(String(event.target));
  })
  .on('complete', function() {
    console.log('Fastest is ' + this.filter('fastest').map('name'));
  })
  .run();


tests/test.spec.js
/*!
* Copyright (c) 2020-2022 Digital Bazaar, Inc. All rights reserved.
*/
import {default as chai} from 'chai';
const should = chai.should();

import {
  base32Encode,
  generateSecret,
  generateToken,
  verify,
  toKeyUri,
  fromKeyUri
} from '../lib/index.js';

describe('totp', () => {
  describe('generateSecret', () => {
    it('should pass with default params', async () => {
      const result = await generateSecret();
      should.exist(result);
      result.should.be.an('object');
      result.should.include.keys(['secret', 'algorithm']);
      result.algorithm.should.equal('SHA-1');
      (result.secret instanceof Uint8Array).should.equal(true);
      result.secret.length.should.equal(20);
    });
    it('should pass with SHA-1', async () => {
      const algorithm = 'SHA-1';
      const result = await generateSecret({algorithm});
      should.exist(result);
      result.should.be.an('object');
      result.should.include.keys(['secret', 'algorithm']);
      result.algorithm.should.equal('SHA-1');
      (result.secret instanceof Uint8Array).should.equal(true);
      result.secret.length.should.equal(20);
    });
    it('should pass with SHA-256', async () => {
      const algorithm = 'SHA-256';
      const result = await generateSecret({algorithm});
      should.exist(result);
      result.should.be.an('object');
      result.should.include.keys(['secret', 'algorithm']);
      result.algorithm.should.equal('SHA-256');
      (result.secret instanceof Uint8Array).should.equal(true);
      result.secret.length.should.equal(32);
    });
    it('should pass with SHA-512', async () => {
      const algorithm = 'SHA-512';
      const result = await generateSecret({algorithm});
      should.exist(result);
      result.should.be.an('object');
      result.should.include.keys(['secret', 'algorithm']);
      result.algorithm.should.equal('SHA-512');
      (result.secret instanceof Uint8Array).should.equal(true);
      result.secret.length.should.equal(64);
    });
    it('should fail with unsupported algorithm', async () => {
      let result;
      let err;
      try {
        result = await generateSecret({algorithm: 'BLAKE2b'});
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('Error');
      err.message.should.equal('Unsupported hash algorithm "BLAKE2b".');
    });
  });

  describe('generateToken', () => {
    it('should pass with default params', async () => {
      const {secret} = await generateSecret();
      const result = await generateToken({secret});
      should.exist(result);
      result.should.be.a('object');
      result.should.include.keys(['token', 'algorithm', 'digits', 'period']);
      result.algorithm.should.equal('SHA-1');
      result.digits.should.equal(6);
      result.period.should.equal(30);
    });
    it('should pass with custom params', async () => {
      const {secret} = await generateSecret();
      const result = await generateToken({
        secret,
        algorithm: 'SHA-256',
        digits: 7,
        period: 45
      });
      should.exist(result);
      result.should.be.an('object');
      result.should.include.keys(['token', 'algorithm', 'digits', 'period']);
      result.algorithm.should.equal('SHA-256');
      result.digits.should.equal(7);
      result.period.should.equal(45);
    });
    it('should fail with no secret', async () => {
      let result;
      let err;
      try {
        result = await generateToken();
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('TypeError');
      err.message.should.equal(
        '"secret" must be Uint8Array or base32-encoded string.');
    });
    it('should fail with digits > 10', async () => {
      const {secret} = await generateSecret();
      let result;
      let err;
      try {
        result = await generateToken({secret, digits: 11});
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('Error');
      err.message.should.equal('"digits" must be an integer > 0 and <= 10.');
    });
    it('should fail with digits < 0', async () => {
      const {secret} = await generateSecret();
      let result;
      let err;
      try {
        result = await generateToken({secret, digits: -1});
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('Error');
      err.message.should.equal('"digits" must be an integer > 0 and <= 10.');
    });
    it('should fail with unsupported algorithm', async () => {
      const {secret} = await generateSecret();
      let result;
      let err;
      try {
        result = await generateToken({secret, algorithm: 'BLAKE2b'});
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('Error');
      err.message.should.equal('Unsupported hash algorithm "BLAKE2b".');
    });
  });

  describe('verify', () => {
    it('should verify with default params', async () => {
      const {secret} = await generateSecret();
      const {token} = await generateToken({secret});
      const result = await verify({token, secret});
      should.exist(result);
      result.should.equal(true);
    });
    it('should not verify with default params', async () => {
      const {secret} = await generateSecret();
      let {token} = await generateToken({secret});
      token = (token[0] === '1' ? '0' : '1') + token.slice(1);
      const result = await verify({token, secret, delta: 0});
      should.exist(result);
      result.should.equal(false);
    });
    it('should verify with SHA-1', async () => {
      const algorithm = 'SHA-1';
      const {secret} = await generateSecret();
      const {token} = await generateToken({secret, algorithm});
      const result = await verify({token, secret, algorithm});
      should.exist(result);
      result.should.equal(true);
    });
    it('should not verify with SHA-1', async () => {
      const algorithm = 'SHA-1';
      const {secret} = await generateSecret();
      let {token} = await generateToken({secret, algorithm});
      token = (token[0] === '1' ? '0' : '1') + token.slice(1);
      const result = await verify({token, secret, algorithm, delta: 0});
      should.exist(result);
      result.should.equal(false);
    });
    it('should verify with SHA-256', async () => {
      const algorithm = 'SHA-256';
      const {secret} = await generateSecret();
      const {token} = await generateToken({secret, algorithm});
      const result = await verify({token, secret, algorithm});
      should.exist(result);
      result.should.equal(true);
    });
    it('should not verify with SHA-256', async () => {
      const algorithm = 'SHA-256';
      const {secret} = await generateSecret();
      let {token} = await generateToken({secret, algorithm});
      token = (token[0] === '1' ? '0' : '1') + token.slice(1);
      const result = await verify({token, secret, algorithm, delta: 0});
      should.exist(result);
      result.should.equal(false);
    });
    it('should verify with SHA-512', async () => {
      const algorithm = 'SHA-512';
      const {secret} = await generateSecret();
      const {token} = await generateToken({secret, algorithm});
      const result = await verify({token, secret, algorithm});
      should.exist(result);
      result.should.equal(true);
    });
    it('should not verify with SHA-512', async () => {
      const algorithm = 'SHA-512';
      const {secret} = await generateSecret();
      let {token} = await generateToken({secret, algorithm});
      token = (token[0] === '1' ? '0' : '1') + token.slice(1);
      const result = await verify({token, secret, algorithm, delta: 0});
      should.exist(result);
      result.should.equal(false);
    });
    it('should fail with no token', async () => {
      const {secret} = await generateSecret();
      let result;
      let err;
      try {
        result = await verify({secret});
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('TypeError');
      err.message.should.equal('"token" must be a string.');
    });
    it('should fail with token length === 0', async () => {
      const {secret} = await generateSecret();
      let result;
      let err;
      try {
        result = await verify({token: '', secret});
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('Error');
      err.message.should.equal('"token" length must be > 0 and <= 10.');
    });
    it('should fail with token length > 10', async () => {
      const {secret} = await generateSecret();
      let result;
      let err;
      try {
        result = await verify({token: '01234567891', secret});
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('Error');
      err.message.should.equal('"token" length must be > 0 and <= 10.');
    });
    it('should fail with no secret', async () => {
      const {secret} = await generateSecret();
      const {token} = await generateToken({secret});
      let result;
      let err;
      try {
        result = await verify({token});
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('TypeError');
      err.message.should.equal(
        '"secret" must be Uint8Array or base32-encoded string.');
    });
  });

  describe('toKeyUri', () => {
    it('should pass', async () => {
      const params = await generateSecret();
      const uri = toKeyUri({
        ...params,
        accountname: 'test@site.example'
      });
      const expected = 'otpauth://totp/test@site.example?secret=' +
        base32Encode(params.secret);
      expected.should.equal(uri);
    });
    it('should pass with custom params', async () => {
      const uri = toKeyUri({
        accountname: 'test@site.example',
        issuer: 'AnIssuer',
        secret: 'G4NYN7TPBY7ONUNDXIHHW4FVZI',
        algorithm: 'SHA-256',
        digits: 7,
        period: 45
      });
      const expected = 'otpauth://totp/AnIssuer:test@site.example?secret=' +
        'G4NYN7TPBY7ONUNDXIHHW4FVZI&algorithm=SHA256&period=45&digits=7' +
        '&issuer=AnIssuer';
      expected.should.equal(uri);
    });
    it('should fail without an accountname', async () => {
      const params = await generateSecret();
      delete params.secret;
      let result;
      let err;
      try {
        result = toKeyUri({
          ...params
        });
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('TypeError');
      err.message.should.equal('"accountname" must be a non-empty string.');
    });
    it('should fail without a secret', async () => {
      let result;
      let err;
      try {
        result = toKeyUri({
          accountname: 'test@site.example'
        });
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('TypeError');
      err.message.should.equal(
        '"secret" must be Uint8Array or base32-encoded string.');
    });
  });

  describe('fromKeyUri', () => {
    it('should pass', async () => {
      const secret = 'G4NYN7TPBY7ONUNDXIHHW4FVZI';
      const uri = `otpauth://totp/test@site.example?secret=${secret}`;
      const result = fromKeyUri({uri});
      should.exist(result);
      result.should.be.an('object');
      result.should.include.keys([
        'secret', 'algorithm', 'digits', 'period', 'accountname'
      ]);
      result.accountname.should.equal('test@site.example');
      result.algorithm.should.equal('SHA-1');
      result.digits.should.equal(6);
      result.period.should.equal(30);
      result.secret.should.equal(secret);
    });
    it('should pass with custom params', async () => {
      const secret = 'G4NYN7TPBY7ONUNDXIHHW4FVZI';
      const uri = 'otpauth://totp/AnIssuer:test@site.example?secret=' +
        'G4NYN7TPBY7ONUNDXIHHW4FVZI&algorithm=SHA256&period=45&digits=7' +
        '&issuer=AnIssuer';
      const result = fromKeyUri({uri});
      should.exist(result);
      result.should.be.an('object');
      result.should.include.keys([
        'secret', 'algorithm', 'digits', 'period', 'accountname', 'issuer'
      ]);
      result.accountname.should.equal('test@site.example');
      result.issuer.should.equal('AnIssuer');
      result.algorithm.should.equal('SHA-256');
      result.digits.should.equal(7);
      result.period.should.equal(45);
      result.secret.should.equal(secret);
    });
    it('should fail with unknown type', async () => {
      let result;
      let err;
      try {
        const uri = 'otpauth://hotp?secret=algorithm=BLAKE2b';
        result = fromKeyUri({uri});
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('Error');
      err.message.should.equal('Unknown supported type "hotp".');
    });
    it('should fail with unknown hash algorithm', async () => {
      let result;
      let err;
      try {
        const uri = 'otpauth://totp?algorithm=BLAKE2b';
        result = fromKeyUri({uri});
      } catch(e) {
        err = e;
      }
      should.not.exist(result);
      should.exist(err);
      err.name.should.equal('Error');
      err.message.should.equal('Unsupported hash algorithm "BLAKE2b".');
    });
  });

  /* The test token shared secret uses the ASCII string value
  "12345678901234567890".  With Time Step X = 30, and the Unix epoch as
  the initial value to count time steps, where T0 = 0, the TOTP
  algorithm will display the following values for specified modes and
  timestamps.

  +-------------+--------------+------------------+----------+--------+
  |  Time (sec) |   UTC Time   | Value of T (hex) |   TOTP   |  Mode  |
  +-------------+--------------+------------------+----------+--------+
  |      59     |  1970-01-01  | 0000000000000001 | 94287082 |  SHA1  |
  |             |   00:00:59   |                  |          |        |
  |      59     |  1970-01-01  | 0000000000000001 | 46119246 | SHA256 |
  |             |   00:00:59   |                  |          |        |
  |      59     |  1970-01-01  | 0000000000000001 | 90693936 | SHA512 |
  |             |   00:00:59   |                  |          |        |
  |  1111111109 |  2005-03-18  | 00000000023523EC | 07081804 |  SHA1  |
  |             |   01:58:29   |                  |          |        |
  |  1111111109 |  2005-03-18  | 00000000023523EC | 68084774 | SHA256 |
  |             |   01:58:29   |                  |          |        |
  |  1111111109 |  2005-03-18  | 00000000023523EC | 25091201 | SHA512 |
  |             |   01:58:29   |                  |          |        |
  |  1111111111 |  2005-03-18  | 00000000023523ED | 14050471 |  SHA1  |
  |             |   01:58:31   |                  |          |        |
  |  1111111111 |  2005-03-18  | 00000000023523ED | 67062674 | SHA256 |
  |             |   01:58:31   |                  |          |        |
  |  1111111111 |  2005-03-18  | 00000000023523ED | 99943326 | SHA512 |
  |             |   01:58:31   |                  |          |        |
  |  1234567890 |  2009-02-13  | 000000000273EF07 | 89005924 |  SHA1  |
  |             |   23:31:30   |                  |          |        |
  |  1234567890 |  2009-02-13  | 000000000273EF07 | 91819424 | SHA256 |
  |             |   23:31:30   |                  |          |        |
  |  1234567890 |  2009-02-13  | 000000000273EF07 | 93441116 | SHA512 |
  |             |   23:31:30   |                  |          |        |
  |  2000000000 |  2033-05-18  | 0000000003F940AA | 69279037 |  SHA1  |
  |             |   03:33:20   |                  |          |        |
  |  2000000000 |  2033-05-18  | 0000000003F940AA | 90698825 | SHA256 |
  |             |   03:33:20   |                  |          |        |
  |  2000000000 |  2033-05-18  | 0000000003F940AA | 38618901 | SHA512 |
  |             |   03:33:20   |                  |          |        |
  | 20000000000 |  2603-10-11  | 0000000027BC86AA | 65353130 |  SHA1  |
  |             |   11:33:20   |                  |          |        |
  | 20000000000 |  2603-10-11  | 0000000027BC86AA | 77737706 | SHA256 |
  |             |   11:33:20   |                  |          |        |
  | 20000000000 |  2603-10-11  | 0000000027BC86AA | 47863826 | SHA512 |
  |             |   11:33:20   |                  |          |        |
  +-------------+--------------+------------------+----------+--------+ */
  describe('RFC 6238 Test Vectors', () => {
    // secret used with all test vectors
    const secret = new TextEncoder().encode('12345678901234567890');
    const vectors = [
      {time: 59, expectedToken: '94287082', algorithm: 'SHA-1'},
      {time: 59, expectedToken: '46119246', algorithm: 'SHA-256'},
      {time: 59, expectedToken: '90693936', algorithm: 'SHA-512'},
      {time: 1111111109, expectedToken: '07081804', algorithm: 'SHA-1'},
      {time: 1111111109, expectedToken: '68084774', algorithm: 'SHA-256'},
      {time: 1111111109, expectedToken: '25091201', algorithm: 'SHA-512'},
      {time: 1111111111, expectedToken: '14050471', algorithm: 'SHA-1'},
      {time: 1111111111, expectedToken: '67062674', algorithm: 'SHA-256'},
      {time: 1111111111, expectedToken: '99943326', algorithm: 'SHA-512'},
      {time: 1234567890, expectedToken: '89005924', algorithm: 'SHA-1'},
      {time: 1234567890, expectedToken: '91819424', algorithm: 'SHA-256'},
      {time: 1234567890, expectedToken: '93441116', algorithm: 'SHA-512'},
      {time: 2000000000, expectedToken: '69279037', algorithm: 'SHA-1'},
      {time: 2000000000, expectedToken: '90698825', algorithm: 'SHA-256'},
      {time: 2000000000, expectedToken: '38618901', algorithm: 'SHA-512'},
      {time: 20000000000, expectedToken: '65353130', algorithm: 'SHA-1'},
      {time: 20000000000, expectedToken: '77737706', algorithm: 'SHA-256'},
      {time: 20000000000, expectedToken: '47863826', algorithm: 'SHA-512'}
    ];

    for(const [i, vector] of vectors.entries()) {
      it(`should pass test vector ${i + 1}`, async () => {
        const {time, expectedToken, algorithm} = vector;
        const now = time * 1000;
        const {token} = await generateToken({
          secret, algorithm, digits: expectedToken.length, now
        });
        token.should.equal(expectedToken);
      });
    }
  });
});


lib/hmac.js
/*!
 * Copyright (c) 2022 Digital Bazaar, Inc. All rights reserved.
 */
import {crypto} from './crypto.js';

export const SUPPORTED_HASHES = ['SHA-1', 'SHA-256', 'SHA-512'];

// TOTP secrets are padded to digest size by repeating the secret
// SHA-1: 20 bytes
// SHA-256: 32 bytes
// SHA-512: 64 bytes
export const MIN_SECRET_SIZE = new Map([
  ['SHA-1', 20],
  ['SHA-256', 32],
  ['SHA-512', 64],
]);

const EXTRACTABLE = false;
const KEY_USAGE = ['sign'];

const subtle = _getCryptoSubtle();
const getRandomValues = _getRandomValues();

export async function generateSecret({size = 20} = {}) {
  return getRandomValues(new Uint8Array(size));
}

// TOTP hmac will pad secret to algorithm size per RFC 6238
export async function hmac({algorithm, secret, data}) {
  if(!SUPPORTED_HASHES.includes(algorithm)) {
    throw new Error(`Unsupported hash algorithm "${algorithm}".`);
  }
  if(typeof data === 'string') {
    data = new TextEncoder().encode(data);
  }
  // pad secret according to RFC 6238
  secret = _padSecret({algorithm, secret});
  const hmacAlgorithm = {
    name: 'HMAC', hash: {name: algorithm}, length: secret.length * 8
  };
  const key = await subtle.importKey(
    'raw', secret, hmacAlgorithm, EXTRACTABLE, KEY_USAGE);
  return new Uint8Array(await subtle.sign(key.algorithm, key, data));
}

// TOTP secrets are padded by repeating the secret if insufficient in length
function _padSecret({algorithm, secret}) {
  const minSize = MIN_SECRET_SIZE.get(algorithm);
  if(secret.length >= minSize) {
    return secret;
  }
  const padded = new Uint8Array(minSize);
  let length = 0;
  while(length < minSize) {
    const remaining = minSize - length;
    if(secret.length > remaining) {
      padded.set(secret.subarray(0, remaining), length);
      break;
    }
    padded.set(secret, length);
    length += secret.length;
  }
  return padded;
}

function _getCryptoSubtle() {
  const subtle = crypto?.webcrypto?.subtle ?? crypto?.subtle ?? {};
  if(subtle.importKey) {
    return subtle;
  }

  // local node.js 14.x polyfill supports just `hmac`
  subtle.importKey = async function importKey(format, secret, algorithm) {
    return {format, secret, algorithm};
  };
  subtle.sign = async function sign(algorithm, key, data) {
    const {secret} = key;
    const hash = algorithm.hash.name.toLowerCase().replace(/-/g, '');
    const hmac = crypto.createHmac(hash, secret).update(data).digest();
    return new Uint8Array(hmac);
  };

  return subtle;
}

function _getRandomValues() {
  if(crypto?.getRandomValues) {
    return crypto.getRandomValues.bind(crypto);
  }

  if(crypto?.webcrypto?.getRandomValues) {
    return crypto.webcrypto.getRandomValues.bind(crypto.webcrypto);
  }

  if(crypto.randomFill) {
    return async function randomFill(x) {
      return new Promise((resolve, reject) =>
        crypto.randomFill(
          x, (err, result) => err ? reject(err) : resolve(result)));
    };
  }

  throw new Error(
    'Web Crypto "getRandomValues" or "crypto.randomFill" required.');
}


lib/totp.js
/*!
 * Copyright (c) 2022 Digital Bazaar, Inc. All rights reserved.
 */
import * as hmac from './hmac.js';
import {default as _base32Encode} from 'base32-encode';
import {default as _base32Decode} from 'base32-decode';

// Note: RFC4648 is an alias for RFC3548 (it is the updated RFC)
const BASE32_VARIANT = 'RFC4648';

/**
 * Generates a new TOTP secret.
 *
 * @param {object} options - The options to use.
 * @param {string} [options.algorithm='SHA-1'] - The hash algorithm to use;
 *   supported algorithms are: 'SHA-1', 'SHA-256', and 'SHA-512'.
 *
 * @returns {Promise<string>} - Resolves to the TOTP token.
 */
export async function generateSecret({algorithm = 'SHA-1'} = {}) {
  const size = hmac.MIN_SECRET_SIZE.get(algorithm);
  if(size === undefined) {
    throw new Error(`Unsupported hash algorithm "${algorithm}".`);
  }
  const secret = await hmac.generateSecret({size});
  return {algorithm, secret};
}

/**
 * Generates a new TOTP token to be verified by an authenticator.
 *
 * @param {object} options - The options to use.
 * @param {Uint8Array|string} options.secret - The secret to use, as a byte
 *   array or a base32-encoded string.
 * @param {string} [options.algorithm='SHA-1'] - The hash algorithm to use;
 *   supported algorithms are: 'SHA-1', 'SHA-256', and 'SHA-512'.
 * @param {number} [options.digits=6] - The number of digits to include
 *   in the token.
 * @param {number} [options.period=30] - The number of seconds in each TOTP
 *   period; used to determine the number of TOTP steps since the current
 *   time.
 * @param {number} [options.now=Date.now()] - The current time; this should
 *   only be used for testing purposes to avoid misuse.
 *
 * @returns {Promise<object>} - Resolves to the params and TOTP token.
 */
export async function generateToken({
  secret, algorithm = 'SHA-1', digits = 6, period = 30, now = Date.now()
} = {}) {
  if(typeof digits !== 'number' || !(digits > 0 && digits <= 10)) {
    throw new Error('"digits" must be an integer > 0 and <= 10.');
  }
  secret = _decodeSecret({secret});

  // calculate the number of steps since time=0
  const steps = _getCurrentSteps({now, period});
  const token = await _generateToken({steps, algorithm, secret, digits});
  return {token, algorithm, digits, period};
}

/**
 * Verifies whether a TOTP token is authentic.
 *
 * @param {object} options - The options to use.
 * @param {string} options.token - The TOTP token to verify.
 * @param {Uint8Array|string} options.secret - The secret to use, as a byte
 *   array or a base32-encoded string.
 * @param {string} [options.algorithm='SHA-1'] - The hash algorithm to use;
 *   supported algorithms are: 'SHA-1', 'SHA-256', and 'SHA-512'.
 * @param {number} [options.period=30] - The number of seconds in each TOTP
 *   period; used to determine the number of TOTP steps since the current time.
 * @param {number} [options.delta=1] - The maximum number of steps (each step
 *   being one period in length) to allow in either direction around the
 *   current time when verifying.
 * @param {number} [options.now=Date.now()] - The current time; this should
 *   only be used for testing purposes to avoid misuse.
 *
 * @returns {Promise<string>} - Resolves to the TOTP token.
 */
export async function verify({
  token, secret, algorithm = 'SHA-1', period = 30, delta = 1, now = Date.now()
} = {}) {
  if(typeof token !== 'string') {
    throw new TypeError('"token" must be a string.');
  }
  if(!(token.length > 0 && token.length <= 10)) {
    throw new Error('"token" length must be > 0 and <= 10.');
  }
  if(typeof delta !== 'number' || !(delta >= 0 && delta <= 10)) {
    throw new Error('"delta" must be an integer >= 0 and <= 10.');
  }

  secret = _decodeSecret({secret});
  const digits = token.length;

  // hmac the token prior to comparison to mitigate timing attacks
  const verifyValue = await hmac.hmac({algorithm, secret, data: token});

  // calculate the number of steps since time=0
  const currentSteps = _getCurrentSteps({now, period});

  // note: no effort made to randomize which tokens are checked first to
  // thwart timing attacks, etc.
  const promises = [];
  for(let i = -delta; i <= delta; ++i) {
    const steps = Math.max(0, currentSteps + i);
    promises.push(_generateVerifyValue({steps, algorithm, secret, digits}));
  }
  const candidates = await Promise.all(promises);

  // return true if any candidate matches
  return candidates.some(c => {
    for(let i = 0; i < c.length; ++i) {
      if(c[i] !== verifyValue[i]) {
        return false;
      }
    }
    return true;
  });
}

// https://github.com/google/google-authenticator/wiki/Key-Uri-Format
export function toKeyUri({
  accountname, issuer, secret, algorithm = 'SHA-1', period = 30, digits = 6
} = {}) {
  if(!(typeof accountname === 'string' && accountname.length > 0)) {
    throw new TypeError('"accountname" must be a non-empty string.');
  }
  if(!(typeof secret === 'string' || secret instanceof Uint8Array)) {
    throw new TypeError(
      '"secret" must be Uint8Array or base32-encoded string.');
  }
  if(issuer && issuer.includes(':')) {
    issuer = encodeURIComponent(issuer);
  }
  if(accountname.includes(':')) {
    accountname = encodeURIComponent(accountname);
  }
  const label = issuer ? `${issuer}:${accountname}` : accountname;
  if(typeof secret !== 'string') {
    secret = base32Encode(secret);
  }
  const params = {secret};
  // add non-default params
  if(algorithm !== 'SHA-1') {
    params.algorithm = algorithm.replace(/-/g, '');
  }
  if(period !== 30) {
    params.period = period;
  }
  if(digits !== 6) {
    params.digits = digits;
  }
  if(issuer) {
    params.issuer = issuer;
  }
  const query = new URLSearchParams(params);
  return `otpauth://totp/${label}?${query}`;
}

export function fromKeyUri({uri} = {}) {
  const parsed = new URL(uri);
  const {protocol, searchParams} = parsed;
  if(protocol !== 'otpauth:') {
    throw new Error(`Unknown protocol "${protocol}".`);
  }
  // `host` and `pathname` may be parsed differently on different platforms;
  // normalize them
  let {host, pathname} = parsed;
  if(host === '') {
    // in this case see if `pathname` is parsed as `//totp/...`
    if(pathname.startsWith('//')) {
      const nextSlash = pathname.indexOf('/', 2);
      host = pathname.substring(2, nextSlash === -1 ? undefined : nextSlash);
      pathname = pathname.slice(2 + host.length);
    }
  }
  if(host !== 'totp') {
    throw new Error(`Unknown supported type "${host}".`);
  }

  const label = pathname.slice(1);
  const {
    secret, algorithm = 'SHA1', period = 30, digits = 6, issuer
  } = Object.fromEntries(searchParams);

  const [issuerPrefix, accountname] = label.split(':');
  if(!algorithm.startsWith('SHA')) {
    throw new Error(`Unsupported hash algorithm "${algorithm}".`);
  }

  return {
    type: 'totp',
    label,
    issuer,
    accountname: decodeURIComponent(accountname || issuerPrefix),
    secret,
    algorithm: `SHA-${algorithm.slice(3)}`,
    period: parseInt(period, 10),
    digits: parseInt(digits, 10)
  };
}

export function base32Encode(data) {
  return _base32Encode(data, BASE32_VARIANT, {padding: false});
}

export function base32Decode(string) {
  const length = Math.floor(string.length * 5 / 8);
  return new Uint8Array(_base32Decode(string, BASE32_VARIANT), 0, length);
}

function _decodeSecret({secret}) {
  if(typeof secret === 'string') {
    // presume base32 encoding
    return base32Decode(secret);
  }
  if(secret instanceof Uint8Array) {
    return secret;
  }
  throw new TypeError('"secret" must be Uint8Array or base32-encoded string.');
}

// returns a 31-bit uint32 'dynamic binary code' from an hmac hash input
function _dynamicTruncation({hash}) {
  // according to RFC 4226, next we find an offset into `hash` by using the
  // last 4 bits of the last byte of `hash`
  const offset = hash[hash.length - 1] & 0x0f;

  // using the offset, we read a 31-bit uint32 (4 bytes with the most
  // significant byte masked with 0x7f; note that since the offset is a 4-bit
  // size it will be <= 16, which will not overrun the min hash size of 20
  // bytes)
  // mask first byte with 0x7f and then read integer
  hash[offset] &= 0x7f;
  const dv = new DataView(hash.buffer, hash.byteOffset, hash.length);
  return dv.getUint32(offset);
}

// generates values for verification using the double hmac mechanism to
// avoid timing attacks
async function _generateVerifyValue({steps, algorithm, secret, digits}) {
  const token = await _generateToken({steps, algorithm, secret, digits});
  return hmac.hmac({algorithm, secret, data: token});
}

async function _generateToken({steps, algorithm, secret, digits}) {
  // steps function as an 8-byte HOTP counter value per RFC 6238 and RFC 4226;
  // ensure to zero-fill to 16 hex digits to get an 8 byte value
  const counter = _hexToUint8Array(steps.toString(16).padStart(16, '0'));
  const hash = await hmac.hmac({algorithm, secret, data: counter});

  // get "dynamic binary code" per RFC 4226
  const dbc = _dynamicTruncation({hash});

  // token is `dbc % 10^<digits>`; i.e., expressing `dbc` as a zero-filled,
  // 10 digit decimal number and taking the last `digits`
  const dbcDec = dbc.toString(10).padStart(10, '0');
  return dbcDec.slice(dbcDec.length - digits);
}

function _getCurrentSteps({now, period}) {
  // convert time to seconds
  const secs = Math.floor(now / 1000);

  // calculate the number of steps since time=0
  return Math.floor(secs / period);
}

function _hexToUint8Array(hex) {
  return new Uint8Array(hex.match(/.{1,2}/g).map(byte => parseInt(byte, 16)));
}
